pkg/storegateway/indexheader/lazy_binary_reader_test.go

```diff
@@ -294,6 +297,71 @@ func testLazyBinaryReader(t *testing.T, bkt objstore.BucketReader, dir string, i
 		return NewStreamBinaryReader(ctx, logger, bkt, dir, id, 3, NewStreamBinaryReaderMetrics(nil), Config{})
 	}
 
-	reader, err := NewLazyBinaryReader(ctx, factory, logger, bkt, dir, id, NewLazyBinaryReaderMetrics(nil), nil)
+	reader, err := NewLazyBinaryReader(ctx, factory, logger, bkt, dir, id, NewLazyBinaryReaderMetrics(nil), nil, gate.NewNoop())
 	test(t, reader, err)
 }
+
+// Tests if LazyBinaryReader blocks concurrent loads such that it doesn't pass the configured maximum
+func TestLazyBinaryReader_ShouldBlockMaxConcurrency(t *testing.T) {
+	ctx := context.Background()
+
+	tmpDir := filepath.Join(t.TempDir(), "test-indexheader")
+	bkt, err := filesystem.NewBucket(filepath.Join(tmpDir, "bkt"))
+	require.NoError(t, err)
+	t.Cleanup(func() { require.NoError(t, bkt.Close()) })
+
+	// Create block.
+	blockID, err := block.CreateBlock(ctx, tmpDir, []labels.Labels{
+		labels.FromStrings("a", "1"),
+		labels.FromStrings("a", "2"),
+		labels.FromStrings("a", "3"),
+	}, 100, 0, 1000, labels.FromStrings("ext1", "1"))
+	require.NoError(t, err)
+	require.NoError(t, block.Upload(ctx, log.NewNopLogger(), bkt, filepath.Join(tmpDir, blockID.String()), nil))
+
+	logger := log.NewNopLogger()
+
+	const (
+		numLazyReader          = 20 // 20 lazy readers
+		maxLazyLoadConcurrency = 3  // maxConcurrency = 3
+	)
+
+	total := atomic.NewUint32(0)
+	inflight := atomic.NewUint32(0)
+
+	factory := func() (Reader, error) {
+		inflight.Inc()
+		require.LessOrEqual(t, inflight.Load(), uint32(maxLazyLoadConcurrency))
+		total.Inc()
+
+		time.Sleep(3 * time.Second)
+
+		inflight.Dec()
+
+		return nil, errors.New("oh no")
+	}
+
+	var lazyReaders [numLazyReader]*LazyBinaryReader
+	lazyLoadingGate := gate.NewInstrumented(prometheus.NewRegistry(), maxLazyLoadConcurrency, gate.NewBlocking(maxLazyLoadConcurrency))
+
+	for i := 0; i < numLazyReader; i++ {
+		lazyReaders[i], err = NewLazyBinaryReader(ctx, factory, logger, bkt, tmpDir, blockID, NewLazyBinaryReaderMetrics(nil), nil, lazyLoadingGate)
+		require.NoError(t, err)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(numLazyReader)
+
+	// Attempt to concurrently load 20 index-headers.
+	for i := 0; i < numLazyReader; i++ {
+		index := i
+		go func() {
+			_, err := lazyReaders[index].IndexVersion()
+			require.Error(t, err)
```

Comment:
Can it be a bit more specific what error we're expecting? You can add a sentinel value that you return from the individual goroutines and then use `require.ErrorIs` here to make sure it's the same error and not something unrelated.